### PR TITLE
fix(kommodity-cluster): always render autobootstrap ExtensionServiceConfig to prevent etcd split-brain [PLA-6531]

### PR DESCRIPTION
**Pull request by andreaswachs**
### Problem

When provisioning a new cluster with 3 control-plane replicas, etcd split-brain occurs: multiple control-plane nodes form independent etcd clusters instead of one joining another. The cluster never reaches a healthy state and requires manual intervention or full teardown.

### Root cause

The `kommodity-autobootstrap` Talos extension is baked into every Talos image unconditionally (via `DEFAULT_EXTENSIONS` in the `talos-cloud-image.yml` GitHub workflow). The extension starts automatically on every control-plane node after `apid` and network are up.

For `public: true` clusters, the chart omitted the `ExtensionServiceConfig` (controlplane.yaml line 86), so the extension ran with compiled-in defaults (`quorumNodes=1`). Any single control-plane node could bootstrap independently after a 10-second delay. When multiple nodes couldn't discover each other during the peer scan (2s timeout per node, 30s interval — too short for cloud VM provisioning variance), each bootstrapped independently, creating separate etcd clusters.

### Fix

1. **Always render the autobootstrap `ExtensionServiceConfig`** — the extension is always baked in, so its config should always be rendered. Removed the `public`/`enabled` conditional from `controlplane.yaml`.

2. **Default `quorumNodes` to `kommodity.controlplane.replicas`** — when `quorumNodes` is null (new default), the template uses the CP replica count. This ensures the extension waits for all control-plane nodes to be discoverable before bootstrapping, preventing split-brain.

3. **Removed dead `$needsStrategicPatches` conditional** — since the autobootstrap config is always rendered, `strategicPatches` is always needed. Simplified to always render.

### Files changed

| File | Change |
|------|--------|
| `templates/talos/controlplane.yaml` | Always render `strategicPatches` + autobootstrap `ExtensionServiceConfig`; pass `replicas` to template; removed dead `$needsStrategicPatches` variable |
| `templates/talos/autobootstrap.yaml` | Accept `replicas` param; use `default $replicas $autoBootstrap.quorumNodes` |
| `values.yaml` | `quorumNodes: 1` → `null`; updated comments |
| `Chart.yaml` | Version bump `0.27.1` → `0.28.0` |
| `tests/talos_provider_test.yaml` | Updated test (now includes for public:true); added 2 tests for quorumNodes defaulting |

### Verification

- 248/248 helm unit tests pass (`make run-helm-unit-tests`)
- `git diff --check` clean
- Rendered output verified: `QUORUM_NODES=3` for 3-replica cluster with `public: true`
- Parallel review subagents: all PASS (correctness, spec compliance, edge cases, ponytail, test coverage)
- Full changeset coherence review: PASS

### Scaleway vs Azure: Network Model & Consequences

| Provider | `public` | Node IPs | Talos API (50000) | Autobootstrap before fix | Autobootstrap after fix |
|----------|----------|----------|-------------------|--------------------------|-------------------------|
| Scaleway | false | Private (Scaleway PN) | Via proxy tunnel | Rendered, quorumNodes=1 | Rendered, quorumNodes=replicas |
| Scaleway | true | **Public** | **Direct dial** | NOT rendered (compiled-in quorumNodes=1) | Rendered, quorumNodes=replicas |
| Azure | false | Private (VNet) | Via proxy tunnel | Rendered, quorumNodes=1 | Rendered, quorumNodes=replicas |
| Azure | true | **Private (VNet)** | **Via proxy tunnel** | **NOT rendered** (compiled-in quorumNodes=1) → **SPLIT-BRAIN** | Rendered, quorumNodes=replicas → **FIXED** |

**Key insight**: Azure `public: true` only makes the K8s API LB public (6443). Node VMs NEVER get public IPs. The Talos API (50000) is never exposed on any LB. Kommodity must always use the talos-cluster-proxy tunnel on Azure. This creates a circular dependency (proxy needs K8s API → K8s API needs etcd → etcd needs bootstrap → bootstrap needs proxy). The autobootstrap extension breaks this cycle by running locally. Without the ExtensionServiceConfig, the extension used `quorumNodes=1` → split-brain.

**Scaleway `public: true`** is different — nodes get public IPs, so Kommodity can reach the Talos API directly. The reconciler's `bootstrapCluster()` can work without the proxy. However, the extension was still baked into the image and ran with `quorumNodes=1`, which was dangerous for multi-replica clusters.

### Upgrade Impact: Chart 0.27.1 → 0.28.0

Upgrading triggers a **full control plane rolling update** for clusters where the rendered `strategicPatches` content changes. This is detected by the Sidero TalosControlPlane controller's `MatchesControlPlaneConfig` (uses `reflect.DeepEqual` on the entire TalosConfigSpec including strategicPatches).

**Note**: `machineSpecsHash` does NOT include strategicPatches, so the infrastructure template name (AzureMachineTemplate/ScalewayMachineTemplate) does NOT change. No new infrastructure template is created.

| Configuration | Old (0.27.1) | New (0.28.0) | Rollout? |
|--------------|-------------|-------------|----------|
| `public: true`, any replicas | No autobootstrap config | Autobootstrap with quorumNodes=replicas | **YES** (new strategicPatches content) |
| `public: false`, replicas=1 | quorumNodes=1 | quorumNodes=1 | **NO** (identical rendered output) |
| `public: false`, replicas=3 | quorumNodes=1 | quorumNodes=3 | **YES** (quorumNodes value changed) |
| `public: false`, replicas=2 (Kubevirt) | quorumNodes=1 | quorumNodes=2 | **YES** (quorumNodes value changed) |
| BYOT (any config) | N/A | N/A | **NO** (uses different template, not affected) |

**Rollout is safe**: `RollingUpdate` with `maxSurge: 1`. New machines join the existing etcd cluster via standard Talos join mechanism. Old machines are gracefully removed one at a time (etcd health check → graceful leave → delete). `quorumNodes` only affects initial bootstrap, not join behavior — the extension exits immediately on already-bootstrapped clusters.

**For `public: true` + `replicas: 1`**: brief downtime during rollout (scale 1→2→1). CAPI waits for new node readiness before deleting old, but operators should be aware.

### Edge cases considered

- **Single-node CP (replicas=1)**: `quorumNodes=1`, single node bootstraps immediately — no split-brain possible
- **Scaling up (1→3)**: New machines join existing etcd, `quorumNodes` only gates initial bootstrap — no impact
- **BYOT provider**: Uses `generateType: init` for bootstrap node — not affected by this change
- **Node failure during bootstrap**: `quorumNodes=replicas` means cluster hangs if one node fails to boot. This is the safe trade-off (hang > split-brain). Operators can set `quorumNodes` explicitly (e.g., 2 for 3 replicas) to tolerate one failure.
- **Azure NSG**: `allow-talos-api-from-vnet` rule only for `public: false`, but Azure's default `AllowVNetInBound` rule covers port 50000 in both modes. Peer discovery works.
- **Azure NAT gateway**: CP subnet doesn't get NAT gateway in `public: true`, but the public Standard LB provides outbound SNAT. CP VMs can reach NTP/internet.
- **Residual race on Scaleway `public: true`**: The reconciler's `bootstrapCluster()` can reach nodes directly (no proxy). Theoretically, it could call `c.Bootstrap()` on a different node than the extension's leader. This is a pre-existing issue in the upstream controller, not introduced by this fix. The fix makes it BETTER: with `quorumNodes=replicas`, the extension waits for quorum instead of bootstrapping immediately with `quorumNodes=1`.
- **Helm upgrade**: Triggers CAPI rolling update for affected clusters; new machines join existing etcd — safe and benign.

### CODE REVIEW

**Comment by andreaswachs** [comment]
## Self-Review: Deployment Config Impact Analysis

### Critical finding: base/values.yaml overrides quorumNodes to 1

The deployments repo (`kommodity/base/values.yaml` line 316) explicitly sets `talos.autoBootstrap.quorumNodes: 1`. This value is inherited by **all** clusters via the sync pipeline (`sync-kommodity-clusters.yml` lines 292-293: `--values base/values.yaml --values <cluster>/values.yaml`).

Since Helm's `default $replicas $autoBootstrap.quorumNodes` returns `1` (the explicit value, not the null default), **the chart fix is ineffective for all existing clusters** — `quorumNodes` remains `1` everywhere.

### Per-cluster impact

| Cluster | Provider | `public` | `replicas` | `quorumNodes` (rendered) | ExtensionServiceConfig before fix | After fix | Rollout? | Split-brain fixed? |
|---------|----------|----------|------------|--------------------------|-----------------------------------|-----------|----------|-------------------|
| dev-ams99 | Azure | `true` | 3 | **1** | NOT rendered (compiled-in 1) | Rendered with 1 | **YES** (new config block) | **NO** |
| dev-par01 | Scaleway | `false` | 3 | **1** | Rendered with 1 | Rendered with 1 | NO | NO |
| dev-par02 | Scaleway | `false` | 3 | **1** | Rendered with 1 | Rendered with 1 | NO | NO |
| dev-par03 | Scaleway | `false` | 3 | **1** | Rendered with 1 | Rendered with 1 | NO | NO |
| dev-par04 | Kubevirt | `false` | 3* | **1** | Rendered with 1 | Rendered with 1 | NO | NO |
| prd-par01 | Scaleway | `false` | 5 | **1** | Rendered with 1 | Rendered with 1 | NO | NO |
| prd-par02 | Scaleway | `false` | 5 | **1** | Rendered with 1 | Rendered with 1 | NO | NO |
| stg-par01 | Scaleway | `false` | 5 | **1** | Rendered with 1 | Rendered with 1 | NO | NO |
| stg-par02 | Scaleway | `false` | 5 | **1** | Rendered with 1 | Rendered with 1 | NO | NO |

*dev-par04 doesn't set `controlplane.replicas` — inherits `3` from base.

### What the chart fix actually does for existing clusters

- **dev-ams99 only**: The ExtensionServiceConfig is now explicitly rendered (was absent because `public: true`). The extension now receives explicit env vars instead of compiled-in defaults. But `quorumNodes` is still `1` — the split-brain vulnerability is **unchanged**. A CP rollout is triggered because the strategicPatches content changes.
- **All other clusters**: Zero change. No rollout. No fix.

### Required action: deployments repo

**Remove `quorumNodes: 1` from `kommodity/base/values.yaml`** (line 316) so the chart's new default (`null` → `replicas`) takes effect. This is a deployment repo change, not a chart change.

After removing the base override:

| Cluster | `quorumNodes` after fix | Split-brain fixed? | Rollout? |
|---------|------------------------|-------------------|----------|
| dev-ams99 (3 replicas) | 3 | **YES** | YES |
| dev-par01-04 (3 replicas) | 3 | **YES** | YES |
| prd-par01-02 (5 replicas) | 5 | **YES** | YES |
| stg-par01-02 (5 replicas) | 5 | **YES** | YES |

**For production/staging (5 replicas)**, consider setting `quorumNodes: 3` per-cluster to tolerate 2 node failures during bootstrap while still preventing split-brain.

### Scaleway `public: false` clusters also at risk

Scaleway clusters with `replicas: 3` or `replicas: 5` and `quorumNodes: 1` are **also vulnerable to split-brain** — not just Azure `public: true` clusters. With `quorumNodes=1`, any single node bootstraps independently after 10s. If nodes can't discover each other during the 2s scan window (Scaleway VM provisioning takes 10-60s), multiple nodes bootstrap independently → split-brain.

### Recommendation

1. **Merge this chart PR** — it correctly fixes the chart's default and always renders the ExtensionServiceConfig.
2. **Open a PR in the deployments repo** to remove `quorumNodes: 1` from `kommodity/base/values.yaml`.
3. Optionally add a chart validation guard: fail/warn when `quorumNodes < 2` and `replicas > 1`.

File: charts/kommodity-cluster/values.yaml [hunk 1]
```diff
@@ -393,12 +392,15 @@ talos:
   #   imageName: kommodity-talos-installer:v1.13.0
   #   repository: ghcr.io/kommodity-io
   autoBootstrap:
-    # autoBootstrap will be enabled regardless of the enabled value if kommodity.network.ipv4.public is set to false
+    # No-op; the extension is baked into every Talos image and its
+    # ExtensionServiceConfig is always rendered with the values below.
     enabled: false
-    # Network discovery frequency
+    # Network Discovery frequency
     scanInterval: 30s
```
[the remark this answers is not in the document]
Comment (pthuriot-corti) [reply charts/kommodity-cluster/values.yaml]: this makes sense I think

**Comment by selcorti** [comment]
Two follow-ups worth considering:

**1. `kommodity.controlplane.replicas` is required but undefaulted**

`kommodity.controlplane.replicas` has no `default` anywhere in the chart (values.yaml ~line 271, referenced bare at controlplane.yaml:19). The new `default $replicas $autoBootstrap.quorumNodes` now leans on it implicitly: with `quorumNodes: null` (the new default) and an unset `replicas`, the template renders `TALOS_AUTO_BOOTSTRAP_QUORUM_NODES=` (empty string).

This isn't a *new* bug — line 19 already breaks (`replicas: <empty>`) without it — but the coupling is now load-bearing for the split-brain fix. A defensive `default 1 $.Values.kommodity.controlplane.replicas` inside the default chain, or a `values.schema.json` requiring `replicas`, would make it robust. Low priority since `replicas` is de-facto mandatory.

**2. No test for the `replicas`-unset + `quorumNodes: null` edge**

The three new tests cover public-renders, null→replicas, and explicit-preserved, but the empty-rendering path (the one the first point above is about) is untested. Worth adding one asserting it doesn't render `QUORUM_NODES=` empty — or better, that it errors. Optional.

File: charts/kommodity-cluster/templates/talos/autobootstrap.yaml [hunk 2]
```diff
@@ -5,14 +5,15 @@ This configures environment variables for the kommodity-autobootstrap extension
...
+{{- $replicas := .replicas -}}
 - |
   apiVersion: v1alpha1
   kind: ExtensionServiceConfig
   name: kommodity-autobootstrap
   environment:
     - LOG_LEVEL={{ $logLevel }}
     - TALOS_AUTO_BOOTSTRAP_SCAN_INTERVAL={{ $autoBootstrap.scanInterval }}
-    - TALOS_AUTO_BOOTSTRAP_QUORUM_NODES={{ $autoBootstrap.quorumNodes }}
+    - TALOS_AUTO_BOOTSTRAP_QUORUM_NODES={{ default $replicas $autoBootstrap.quorumNodes }}
```
Comment (pthuriot-corti) [inline charts/kommodity-cluster/templates/talos/autobootstrap.yaml]: I don't think quorum should be the control plane replicas. If we have 5 CP nodes, quorum is reached at 3 nodes, not 5

**Review by pthuriot-corti — CHANGES_REQUESTED** [review]

File: charts/kommodity-cluster/templates/talos/autobootstrap.yaml (hunk 2 above)
Comment (pthuriot-corti) [reply charts/kommodity-cluster/templates/talos/autobootstrap.yaml]: N = replicas
quorum reached at floor(N/2) + 1.
Maybe we can set that in helm

**Comment by andreaswachs** [comment]
Both points addressed:

**1. Quorum formula**: Changed default from `replicas` to `floor(replicas/2) + 1` (etcd quorum). For 3 replicas → 2, for 5 replicas → 3. This tolerates node failures during bootstrap while still preventing split-brain.

**2. Defensive default for `replicas`**: Added `default 1` in the `controlplane.yaml` include call so an unset `replicas` renders `QUORUM_NODES=1` instead of an empty string.

Also added tests for: 5 replicas → quorum 3, 1 replica → quorum 1, and the existing explicit-override test (quorumNodes: 2 with replicas: 3 still yields 2).

File: charts/kommodity-cluster/values.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (pthuriot-corti) [reply charts/kommodity-cluster/values.yaml]: what about this? It doesn't make sense to still have `talos.autoBootstrap.enabled` if it doesn't do anything

[the remark this answers is not in the document]
Comment (andreaswachs) [reply charts/kommodity-cluster/values.yaml]: ah yeah sorry I missed this. Fair point, will remove it

**Review by pthuriot-corti — APPROVED** [review]
